src/custom/state/swap/extension.ts

```diff
@@ -0,0 +1,159 @@
+    return feeRatioAsCurrency.greaterThan(minimalFeeAsCurrency) ? feeRatioAsCurrency : minimalFeeAsCurrency
+  } else {
+    // One or neither is valid, return that one
+    return minimalFeeAsCurrency || feeRatioAsCurrency
+  }
+}
+
+interface ExtendedTradeParams {
+  exactInTrade?: Trade | null
+  exactOutTrade?: Trade | null
+  typedAmountAsCurrency?: CurrencyAmount
+  feeAsCurrency?: CurrencyAmount
+}
+
+/**
+ * extendExactInTrade
+ * @description takes a Uni ExactIn Trade object and returns a custom one with fee adjusted inputAmount
+ */
+function extendExactInTrade(params: Pick<ExtendedTradeParams, 'exactInTrade' | 'typedAmountAsCurrency'>) {
+  const { exactInTrade, typedAmountAsCurrency } = params
+
+  if (!exactInTrade || !typedAmountAsCurrency) return null
+
+  // We need to iverride the Trade object to use different values as we are intercepting initial inputs
+  // and applying fee. For ExactIn orders, we leave outputAmount as is
+  // and only change inputAmount to show the original entry before fee calculation
+  return {
+    ...exactInTrade,
+    inputAmount: typedAmountAsCurrency,
+    inputAmountWithFees: exactInTrade.inputAmount,
+    minimumAmountOut: exactInTrade.minimumAmountOut,
+    maximumAmountIn: exactInTrade.maximumAmountIn
+  }
+}
+
+/**
+ * extendExactOutTrade
+ * @description takes a Uni ExactOut Trade object and returns a custom one with fee adjusted inputAmount
+ */
+function extendExactOutTrade(params: Pick<ExtendedTradeParams, 'exactOutTrade' | 'feeAsCurrency'>) {
+  const { exactOutTrade, feeAsCurrency } = params
+
+  if (!exactOutTrade || !feeAsCurrency) return null
+
+  const inputAmountWithFee = exactOutTrade.inputAmount.add(feeAsCurrency)
+  // We need to override the Trade object to use different values as we are intercepting initial inputs
+  // and applying fee. For ExactOut orders, we leave inputAmount as is
+  // and only change outputAm to show the original entry before fee calculation
+  return {
+    ...exactOutTrade,
+    inputAmount: inputAmountWithFee,
+    minimumAmountOut: exactOutTrade.minimumAmountOut,
+    maximumAmountIn: exactOutTrade.maximumAmountIn
+  }
+}
+
+interface TradeParams {
+  parsedAmount?: CurrencyAmount
+  inputCurrency?: Currency | null
+  outputCurrency?: Currency | null
+  feeInformation?: FeeInformation
+}
+
+/**
+ * useTradeExactInWithFee
+ * @description wraps useTradeExactIn and returns an extended trade object with the fee adjusted values
+ */
+function useTradeExactInWithFee({ parsedAmount, outputCurrency, feeInformation }: Omit<TradeParams, 'inputCurrency'>) {
+  // Using feeInformation info, determine whether minimalFee greaterThan or lessThan feeRatio * sellAmount
+  let fee: CurrencyAmount | undefined
+  if (parsedAmount && feeInformation) {
+    fee = determineFee({
+      feeInformation,
+      inputAmount: parsedAmount
+    })
+  }
+
+  let inputAmount: CurrencyAmount | undefined = parsedAmount
+  if (parsedAmount && fee) {
```

Comment:
I think it's a bit difficult to follow the logic. 
If you apply the change to `determineFee` so it always return a Fee, then your logic could be simplified to sth like this (i think): 

```ts
function useTradeExactInWithFee({ parsedAmount, outputCurrency, feeInformation }: Omit<TradeParams, 'inputCurrency'>) {
  // Using feeInformation info, determine whether minimalFee greaterThan or lessThan feeRatio * sellAmount
  let inputAmount: CurrencyAmount | undefined
  if (parsedAmount && feeInformation) {
    const fee = determineFee({
      feeInformation,
      inputAmount: parsedAmount
    })
    inputAmount = parsedAmount.subtract(fee)
  }

  // Original Uni trade hook
  const exactInTradeUniswap = useTradeExactIn(inputAmount, outputCurrency ?? undefined)

  return extendExactInTrade({
    exactInTrade: exactInTradeUniswap,
    typedAmountAsCurrency: parsedAmount
  })
}
```

Reply:
the problem here would be i the fee is greater than the input amount, we get a negtive adjusted input amount, which would throw as it's not checked inside `useTradeExactIn` - alternative would be to add the check in there, or add another check one line above `useTradeExactIn` to change the value there again but that is the same as what I currently do. I could also change the `getFeeAmount` function to return `undefined` or `null` if the maths results is lessThanOrEqualTo `zero` for example. Though I think ultimately this is clear enough

Reply:
reworked it a little